quadsim/robots/motors.py

```diff
@@ -1,48 +1,54 @@
-""" Implements models for single DC-DC motors and compositional motor groups.
-"""
+"""Implements models for single DC motors and compositional motor groups."""
 import enum
+from dataclasses import dataclass
 from typing import Optional
 from typing import Sequence
 from typing import Tuple
 from typing import Union
 
 import numpy as np
 
-POSITION_INDEX = 0
-KP_INDEX = 1
-VELOCITY_INDEX = 2
-KD_INDEX = 3
-TORQUE_INDEX = 4
-COMMAND_DIMENSION = 5
+_ARRAY = Sequence[float]
+_FloatOrArray = Union[float, _ARRAY]
 
 
 class MotorControlMode(enum.Enum):
     """Different motor control modes.
-    See the documentation of MotorModel class for details."""
+
+    Right now, 3 motor control modes are supported:
+    - POSITION: performs joint-level PD control.
+    - TORQUE: directly takes in motor torque command
+    - HYBRID: takes in a 5-d tuple (pos, kp, vel, kd, torque), and output
+      torque is a sum of PD torque and additional torque.
+    """
 
     POSITION = 0
     TORQUE = 1
     HYBRID = 2
 
 
-# NOTE: Until functionality is added to MotorModel,
-# it is effectively equivalent to a `dataclass`.
+@dataclass
+class MotorCommand:
+    desired_position: _FloatOrArray = 0
+    kp: _FloatOrArray = 0
+    desired_velocity: _FloatOrArray = 0
+    kd: _FloatOrArray = 0
+    desired_extra_torque: _FloatOrArray = 0
```

Comment:
why is it `desired_extra_torque` and not just `desired_torque`?

Reply:
Technically the structure of the `motor_command` is different for each motor control mode:
- For position, it's just desired_position
- For torque, it's just desired_(extra_)_torque
- For hybrid, it's all 5 used here.
I named it "extra" because in the "hybrid" mode, the actual torque is:
kp * (p - p_des) + kd * (v - v_des) + _extra torque_

I agree the naming could be improved here. Let me know how you think.

Reply:
I see. I think this is kind of a confusing API. I would have gotten stuck on this a while I think if I were doing position control where I specify positions directly, then switched to torque control and assumed I should specify torques directly.

Is there a way to make these symmetric (as in both are either `desired` or `desired_delta`) or does that not really make sense?

Reply:
I can:
1) Rename `desired_extra_torque` to `desired_torque`
2) Create an extra field called `desired_torque` for torque control.
Which way do you prefer?

Reply:
From my understanding, option (1) would not be correct?
So I guess option (2) is better?

Reply:
`MotorCommand`:
* `desired_position`
* `desired_velocity`
* `kp`
* `kd`
* `desired_torque`
* `desired_extra_torque`

Add documentation for the control modes.